src/Utilities/FlowAnalysis/Framework/DataFlow/CacheBasedEquatable.cs

```diff
@@ -11,21 +13,48 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
     internal abstract class CacheBasedEquatable<T> : IEquatable<T>
         where T: class
     {
+        private readonly Lazy<ImmutableArray<int>> _lazyHashCodeParts;
         private readonly Lazy<int> _lazyHashCode;
 
         protected CacheBasedEquatable()
         {
 #pragma warning disable CA2214 // Do not call overridable methods in constructors
-                               // https://github.com/dotnet/roslyn-analyzers/issues/1652
-            _lazyHashCode = new Lazy<int>(() => HashUtilities.Combine(GetType().GetHashCode(), ComputeHashCode()));
+            // https://github.com/dotnet/roslyn-analyzers/issues/1652
+            _lazyHashCodeParts = new Lazy<ImmutableArray<int>>(() => ComputeHashCodeParts());
+            _lazyHashCode = new Lazy<int>(ComputeHashCode);
 #pragma warning restore CA2214 // Do not call overridable methods in constructors
         }
 
-        protected abstract int ComputeHashCode();
+        private int ComputeHashCode() => HashUtilities.Combine(_lazyHashCodeParts.Value, GetType().GetHashCode());
+
+        private ImmutableArray<int> ComputeHashCodeParts()
+        {
+            var builder = ImmutableArray.CreateBuilder<int>();
+            ComputeHashCodeParts(builder);
+            return builder.ToImmutable();
+        }
+
+        protected abstract void ComputeHashCodeParts(ImmutableArray<int>.Builder builder);
 
         public sealed override int GetHashCode() => _lazyHashCode.Value;
         public sealed override bool Equals(object obj) => Equals(obj as T);
-        public bool Equals(T other) => _lazyHashCode.Value == (other as CacheBasedEquatable<T>)?._lazyHashCode.Value;
+        public bool Equals(T other)
+        {
+            // Perform fast equality checks first.
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            var otherEquatable = other as CacheBasedEquatable<T>;
+            if (otherEquatable == null || _lazyHashCode.Value != otherEquatable._lazyHashCode.Value)
+            {
+                return false;
+            }
+
+            // Now perform slow check that compares individual hash code parts sequences.
+            return _lazyHashCodeParts.Value.SequenceEqual(otherEquatable._lazyHashCodeParts.Value);
```

Comment:
I wonder if it'd be worth having a "real" Equals() check in debug builds, to help detect if an individual hash code part had a collision.

Reply:
Good idea, I will do it as one of the follow-up changes after I am done with investigating all the known asserts.